### ISSUE TITLE
Assertion failed defining notation within class

### ISSUE THREAD

**Issue by jwiegley** [coq/coq#16377]
#### Description of the problem

The following code:
```coq
Reserved Notation "\[]" (at level 0).

Class hoare (heap : Type) := {
  hempty : heap -> Prop :=
    fun h => h = h
    where "\[]" := hempty
}.
```
yields the error:
```
Error:
Anomaly
"File "vernac/metasyntax.ml", line 1192, characters 17-23: Assertion failed."
Please report at http://coq.inria.fr/bugs/.
```

#### Coq Version

8.15.2

**Comment by jwiegley** [comment]
And yes, the definition is wrong, I was just eliminating as many terms as possible.

**Comment by SkySkimmer** [comment]
Fixed in 8.18 (#17217)
which means this is a duplicate of #14975
